Use subtopologies in new deployment cookiecutter

moisesmata opened the pull request:
| | |
|:---|:---|
|**_Related Issue(s)_**| nasa/fprime#3841 |
|**_Has Unit Tests (y/n)_**| n |
|**_Documentation Included (y/n)_**| y |

---
## Change Description, fixes nasa/fprime#3841

Introduces support for Fprime core subtopoplogies in the deployment cookiecutter template. Users can choose between legacy deployment architecture and the new modular subtoppoogy approach when generating newFprime deployments. 

Updates:
- cookiecutter.json with `yes/no` choices
- Updated `Main.cpp` to handle different communication configurations
- Updated `topology.fpp` to conditionally include subtopologies (same with `TopologyDefs.hpp`)
- `Topology.cpp` is very stripped down when using subtopologies, (code shifted to phased subtopology code).

moisesmata committed:
Initial refactoring of deployment template to include core subtopologies

moisesmata committed:
Fix BaseID usage in deployment template

moisesmata committed:
Merge branch 'nasa:devel' into update-cookiecutter/subtopology-integration

moisesmata committed:
Updated Packets.fppi template and Main.cpp to work with subtopologies

moisesmata committed:
Cookiecutter.json wording modification

moisesmata committed:
Delete unneeded files, shift away from hard coded constants

moisesmata committed:
Update metadata

check-spelling run (pull_request_target) for update-cookiecutter/subtopology-integration

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata committed:
Fix Readme.md template

moisesmata committed:
Option to include adding log instances added

moisesmata committed:
Changed healthPing name for tlmSend from Chantlm->tlmSend

moisesmata committed:
Update metadata

check-spelling run (pull_request_target) for update-cookiecutter/subtopology-integration

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/cookiecutter.json:
```diff
@@ -1,10 +1,17 @@
+        "use_core_subtopologies": "Use F-Prime core subtopologies (CdhCore, ComFprime/ComCcsds, DataProducts, FileHandling)?",
+        "communication_type": "Select communication subtopology type (only if using core subtopologies)",
+        "com_driver_type": "Select communication driver type",
+        "enable_logging": "Enable ComLoggerTee subtopologies for event and telemetry logging?"
```
Remove this question and make it default.

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/hooks/post_gen_project.py:
```diff
@@ -0,0 +1,20 @@
+import os
```
delete this hook

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/CMakeLists.txt:
```diff
@@ -17,11 +17,20 @@ register_fprime_module(
+{%- if cookiecutter.use_core_subtopologies == "yes" and cookiecutter.enable_logging == "yes" %}
+        # ComLoggerTee subtopology instances for logging
+        {{cookiecutter.__include_path_prefix|replace("/", "_")|replace("\\", "_")|replace(" ", "_")|replace("-", "_")}}{{cookiecutter.deployment_name}}_Top_EventLoggerTee
+        {{cookiecutter.__include_path_prefix|replace("/", "_")|replace("\\", "_")|replace(" ", "_")|replace("-", "_")}}{{cookiecutter.deployment_name}}_Top_TlmLoggerTee
```
Remove the unnecessary replace calls

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/topology.fpp:
```diff
@@ -166,6 +218,53 @@ module {{cookiecutter.deployment_name}} {
+
+      # Rate group 1
+      rateGroupDriver.CycleOut[Ports_RateGroups.rateGroup1] -> rateGroup1.CycleIn
+{%- if cookiecutter.use_core_subtopologies == "yes" %}
```
simplify: always use subtopologies .

moisesmata committed:
Simplified cookiecutter logic, Subtopologies Default, Logging option taken away

moisesmata committed:
Handle UART case in Packets.fppi

moisesmata committed:
Made ComCcsds the default framing option

moisesmata committed:
Update metadata

check-spelling run (pull_request_target) for update-cookiecutter/subtopology-integration

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

thomas-bc committed:
Merge branch 'devel' into update-cookiecutter/subtopology-integration

thomas-bc committed:
Update metadata

check-spelling run (pull_request_target) for update-cookiecutter/subtopology-integration

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata committed:
Fix Formatting for pre_gen_project.py hook

moisesmata committed:
Cookiecutter updates to match new subtopology structure

moisesmata committed:
Followed new subtopology state structure

thomas-bc commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/cookiecutter.json:
```diff
@@ -1,10 +1,13 @@
+
 {
     "deployment_name": "MyDeployment",
+    "communication_type": ["ComCcsds", "ComFprime"],
```
use CCSDS by default.
If we want to keep the option open, we can have `fprime-util new --deployment --com-fprime` ?

thomas-bc commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/instances.fpp:
```diff
@@ -13,118 +13,112 @@ module {{cookiecutter.deployment_name}} {
+  {
+      phase Fpp.ToCpp.Phases.configComponents """
+      if (state.hostname != nullptr && state.port != 0) {
+          {{cookiecutter.deployment_name}}::comDriver.configure(state.hostname, state.port);
```
NO PHASES!!

thomas-bc commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/topology.fpp:
```diff
@@ -118,53 +105,22 @@ module {{cookiecutter.deployment_name}} {
       # Rate group 2
       rateGroupDriver.CycleOut[Ports_RateGroups.rateGroup2] -> rateGroup2.CycleIn
-      rateGroup2.RateGroupMemberOut[0] -> cmdSeq.schedIn
+      rateGroup2.RateGroupMemberOut[0] -> FileHandling.cmdSeq.schedIn
```
take out cmdSeq from filehandling

thomas-bc commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/{{cookiecutter.deployment_name}}Topology.cpp:
```diff
@@ -9,24 +9,16 @@
-
-Svc::ComQueue::QueueConfigurationTable configurationTable;
+// Instantiate a system logger that will handle Fw::Logger::log calls
+Os::Console logger;
```
no logger needed here

thomas-bc committed:
Fix test with ActiveLogger renamed to EventManager

moisesmata committed:
Move cmdSeq to Root Ref, no phaed code

moisesmata committed:
Fix README.md syntax

moisesmata committed:
Update metadata

check-spelling run (pull_request_target) for update-cookiecutter/subtopology-integration

Signed-off-by: check-spelling-bot <check-spelling-bot@users.noreply.github.com>
on-behalf-of: @check-spelling <check-spelling-bot@check-spelling.dev>

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/topology.fpp:
```diff
@@ -9,107 +9,95 @@ module {{cookiecutter.deployment_name}} {
+      CdhCore.cmdDisp.seqCmdStatus -> ComCcsds.fprimeRouter.cmdResponseIn
+      
+      # Command Sequencer
+      cmdSeq.comCmdOut -> CdhCore.cmdDisp.seqCmdBuff
```
remove

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/{{cookiecutter.deployment_name}}Packets.fppi:
```diff
@@ -64,80 +68,102 @@ telemetry packets {{cookiecutter.deployment_name}}Packets {
-  packet Version1 id 6 group 2 {
-    {{cookiecutter.deployment_name}}.version.FrameworkVersion
-    {{cookiecutter.deployment_name}}.version.ProjectVersion
+  packet DataProducts id 21 group 3 {
```
Should we include data products by default on new projects?

bocchino replied:
I think so. Data products are a core feature of F Prime, and we should assume they will be used in a standard project.

moisesmata commented on src/fprime/cookiecutter_templates/cookiecutter-fprime-deployment/{{cookiecutter.deployment_name}}/Top/{{cookiecutter.deployment_name}}Topology.cpp:
```diff
@@ -104,36 +42,22 @@ void configureTopology(const TopologyState& state) {
     if (state.hostname != nullptr && state.port != 0) {
         comDriver.configure(state.hostname, state.port);
     }
+{%- elif cookiecutter.com_driver_type == "TcpServer" %}
```
Incorrect, Client and Server configure should be the same

moisesmata committed:
UnPhased code is similar to what was done in Ref

moisesmata committed:
Fix RefTopology.cpp

moisesmata committed:
Add Enum Headers in TopologyDefs.hpp
